src/idpf.rs

```diff
@@ -135,22 +136,70 @@ where
 /// Trait for values to be programmed into an IDPF.
 ///
 /// Values must form an Abelian group, so that they can be secret-shared, and the group operation
-/// must be represented by [`Add`]. An implementation of [`CoinToss`] must be provided to randomly
-/// select a value using PRG output. Values must be encodable and decodable, without need for a
-/// decoding parameter.
+/// must be represented by [`Add`]. Values must be encodable and decodable, without need for a
+/// decoding parameter. Values can be pseudorandomly generated, with a uniform probability
+/// distribution, from PRG output.
 pub trait IdpfValue:
     Add<Output = Self>
     + AddAssign
     + Sub<Output = Self>
-    + ConditionallySelectable
     + ConditionallyNegatable
-    + CoinToss
     + Encode
     + Decode
     + Sized
 {
+    /// Any run-time parameters needed to produce a value.
+    type Parameters;
+
+    /// Generate a value from PRG output. This must have a uniform probability distribution over
+    /// the PRG output.
+    fn generate<S>(seed_stream: &mut S, parameters: &Self::Parameters) -> Self
+    where
+        S: SeedStream;
+
     /// Returns the additive identity.
-    fn zero() -> Self;
+    fn zero(parameters: &Self::Parameters) -> Self;
+
+    /// Conditionally select between two values in constant time.
+    ///
+    /// This is the same as in [`subtle::ConditionallySelectable`], but without the [`Copy`] bound.
+    fn conditional_select(a: &Self, b: &Self, choice: Choice) -> Self;
+}
+
+impl<F> IdpfValue for F
+where
+    F: FieldElement,
+{
+    type Parameters = ();
+
+    fn generate<S>(seed_stream: &mut S, _: &()) -> Self
+    where
+        S: SeedStream,
+    {
+        // This is analogous to `Prng::get()`, but does not make use of a persistent buffer of
+        // `SeedStream` output.
+        let mut buffer = [0u8; 64];
+        assert!(
+            buffer.len() >= F::ENCODED_SIZE,
+            "field is too big for buffer"
+        );
+        loop {
+            seed_stream.fill(&mut buffer[..F::ENCODED_SIZE]);
+            match Self::try_from_random(&buffer[..F::ENCODED_SIZE]) {
+                Ok(x) => return x,
+                Err(FieldError::ModulusOverflow) => continue,
+                Err(err) => panic!("unexpected error: {err}"),
+            }
```

Comment:
This is copied from the `prng` module, right? It might be useful to deduplicate it as it's a little hairy.

Reply:
Finally, a reason to use `std::ops::ControlFlow` :stuck_out_tongue_closed_eyes:

Reply:
Nice :)